// square/serialization/types/CatalogModifier.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatalogModifier = void 0;
const core = __importStar(require("../../core"));
const ModifierLocationOverrides_1 = require("./ModifierLocationOverrides");
const Money_1 = require("./Money");
exports.CatalogModifier = core.serialization.object({
    name: core.serialization.string().optionalNullable(),
    priceMoney: core.serialization.property("price_money", Money_1.Money.optional()),
    onByDefault: core.serialization.property("on_by_default", core.serialization.boolean().optionalNullable()),
    ordinal: core.serialization.number().optionalNullable(),
    modifierListId: core.serialization.property("modifier_list_id", core.serialization.string().optionalNullable()),
    locationOverrides: core.serialization.property("location_overrides", core.serialization.list(ModifierLocationOverrides_1.ModifierLocationOverrides).optionalNullable()),
    kitchenName: core.serialization.property("kitchen_name", core.serialization.string().optionalNullable()),
    imageId: core.serialization.property("image_id", core.serialization.string().optionalNullable()),
    hiddenOnline: core.serialization.property("hidden_online", core.serialization.boolean().optionalNullable()),
});
