// square/serialization/types/Cube.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cube = void 0;
const core = __importStar(require("../../core"));
const CubeJoin_1 = require("./CubeJoin");
const CubeType_1 = require("./CubeType");
const Dimension_1 = require("./Dimension");
const Folder_1 = require("./Folder");
const Hierarchy_1 = require("./Hierarchy");
const Measure_1 = require("./Measure");
const NestedFolder_1 = require("./NestedFolder");
const Segment_1 = require("./Segment");
exports.Cube = core.serialization.object({
    name: core.serialization.string(),
    title: core.serialization.string().optional(),
    type: CubeType_1.CubeType,
    meta: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    description: core.serialization.string().optional(),
    measures: core.serialization.list(Measure_1.Measure),
    dimensions: core.serialization.list(Dimension_1.Dimension),
    segments: core.serialization.list(Segment_1.Segment),
    joins: core.serialization.list(CubeJoin_1.CubeJoin).optional(),
    folders: core.serialization.list(Folder_1.Folder).optional(),
    nestedFolders: core.serialization.list(NestedFolder_1.NestedFolder).optional(),
    hierarchies: core.serialization.list(Hierarchy_1.Hierarchy).optional(),
});
