// square/serialization/types/TransferOrder.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferOrder = void 0;
const core = __importStar(require("../../core"));
const TransferOrderLine_1 = require("./TransferOrderLine");
const TransferOrderStatus_1 = require("./TransferOrderStatus");
exports.TransferOrder = core.serialization.object({
    id: core.serialization.string().optional(),
    sourceLocationId: core.serialization.property("source_location_id", core.serialization.string().optionalNullable()),
    destinationLocationId: core.serialization.property("destination_location_id", core.serialization.string().optionalNullable()),
    status: TransferOrderStatus_1.TransferOrderStatus.optional(),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.string().optional()),
    expectedAt: core.serialization.property("expected_at", core.serialization.string().optionalNullable()),
    completedAt: core.serialization.property("completed_at", core.serialization.string().optional()),
    notes: core.serialization.string().optionalNullable(),
    trackingNumber: core.serialization.property("tracking_number", core.serialization.string().optionalNullable()),
    createdByTeamMemberId: core.serialization.property("created_by_team_member_id", core.serialization.string().optional()),
    lineItems: core.serialization.property("line_items", core.serialization.list(TransferOrderLine_1.TransferOrderLine).optionalNullable()),
    version: core.serialization.bigint().optional(),
});
