// square/api/resources/transferOrders/client/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferOrdersClient = void 0;
const BaseClient_1 = require("../../../../BaseClient");
const core = __importStar(require("../../../../core"));
const headers_1 = require("../../../../core/headers");
const environments = __importStar(require("../../../../environments"));
const handleNonStatusCodeError_1 = require("../../../../errors/handleNonStatusCodeError");
const errors = __importStar(require("../../../../errors/index"));
const serializers = __importStar(require("../../../../serialization/index"));
class TransferOrdersClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_1.normalizeClientOptionsWithAuth)(options);
    }
    /**
     * Creates a new transfer order in [DRAFT](entity:TransferOrderStatus) status. A transfer order represents the intent
     * to move [CatalogItemVariation](entity:CatalogItemVariation)s from one [Location](entity:Location) to another.
     * The source and destination locations must be different and must belong to your Square account.
     *
     * In [DRAFT](entity:TransferOrderStatus) status, you can:
     * - Add or remove items
     * - Modify quantities
     * - Update shipping information
     * - Delete the entire order via [DeleteTransferOrder](api-endpoint:TransferOrders-DeleteTransferOrder)
     *
     * The request requires source_location_id and destination_location_id.
     * Inventory levels are not affected until the order is started via
     * [StartTransferOrder](api-endpoint:TransferOrders-StartTransferOrder).
     *
     * Common integration points:
     * - Sync with warehouse management systems
     * - Automate regular stock transfers
     * - Initialize transfers from inventory optimization systems
     *
     * Creates a [transfer_order.created](webhook:transfer_order.created) webhook event.
     *
     * @param {Square.CreateTransferOrderRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.create({
     *         idempotencyKey: "65cc0586-3e82-384s-b524-3885cffd52",
     *         transferOrder: {
     *             sourceLocationId: "EXAMPLE_SOURCE_LOCATION_ID_123",
     *             destinationLocationId: "EXAMPLE_DEST_LOCATION_ID_456",
     *             expectedAt: "2025-11-09T05:00:00Z",
     *             notes: "Example transfer order for inventory redistribution between locations",
     *             trackingNumber: "TRACK123456789",
     *             createdByTeamMemberId: "EXAMPLE_TEAM_MEMBER_ID_789",
     *             lineItems: [{
     *                     itemVariationId: "EXAMPLE_ITEM_VARIATION_ID_001",
     *                     quantityOrdered: "5"
     *                 }, {
     *                     itemVariationId: "EXAMPLE_ITEM_VARIATION_ID_002",
     *                     quantityOrdered: "3"
     *                 }]
     *         }
     *     })
     */
    create(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__create(request, requestOptions));
    }
    __create(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, "v2/transfer-orders"),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: serializers.CreateTransferOrderRequest.jsonOrThrow(request, {
                    unrecognizedObjectKeys: "strip",
                    omitUndefined: true,
                }),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.CreateTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v2/transfer-orders");
        });
    }
    /**
     * Searches for transfer orders using filters. Returns a paginated list of matching
     * [TransferOrder](entity:TransferOrder)s sorted by creation date.
     *
     * Common search scenarios:
     * - Find orders for a source [Location](entity:Location)
     * - Find orders for a destination [Location](entity:Location)
     * - Find orders in a particular [TransferOrderStatus](entity:TransferOrderStatus)
     *
     * @param {Square.SearchTransferOrdersRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.search({
     *         query: {
     *             filter: {
     *                 sourceLocationIds: ["EXAMPLE_SOURCE_LOCATION_ID_123"],
     *                 destinationLocationIds: ["EXAMPLE_DEST_LOCATION_ID_456"],
     *                 statuses: ["STARTED", "PARTIALLY_RECEIVED"]
     *             },
     *             sort: {
     *                 field: "UPDATED_AT",
     *                 order: "DESC"
     *             }
     *         },
     *         cursor: "eyJsYXN0X3VwZGF0ZWRfYXQiOjE3NTMxMTg2NjQ4NzN9",
     *         limit: 10
     *     })
     */
    search() {
        return __awaiter(this, arguments, void 0, function* (request = {}, requestOptions) {
            const list = core.HttpResponsePromise.interceptFunction((request) => __awaiter(this, void 0, void 0, function* () {
                var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
                const _authRequest = yield this._options.authProvider.getAuthRequest();
                const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
                const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                    url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, "v2/transfer-orders/search"),
                    method: "POST",
                    headers: _headers,
                    contentType: "application/json",
                    queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                    requestType: "json",
                    body: serializers.SearchTransferOrdersRequest.jsonOrThrow(request, {
                        unrecognizedObjectKeys: "strip",
                        omitUndefined: true,
                    }),
                    timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                    maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                    abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                    fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                    logging: this._options.logging,
                });
                if (_response.ok) {
                    return {
                        data: serializers.SearchTransferOrdersResponse.parseOrThrow(_response.body, {
                            unrecognizedObjectKeys: "passthrough",
                            allowUnrecognizedUnionMembers: true,
                            allowUnrecognizedEnumValues: true,
                            skipValidation: true,
                            breadcrumbsPrefix: ["response"],
                        }),
                        rawResponse: _response.rawResponse,
                    };
                }
                if (_response.error.reason === "status-code") {
                    throw new errors.SquareError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
                }
                return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v2/transfer-orders/search");
            }));
            const dataWithRawResponse = yield list(request).withRawResponse();
            return new core.Page({
                response: dataWithRawResponse.data,
                rawResponse: dataWithRawResponse.rawResponse,
                hasNextPage: (response) => (response === null || response === void 0 ? void 0 : response.cursor) != null && !(typeof (response === null || response === void 0 ? void 0 : response.cursor) === "string" && (response === null || response === void 0 ? void 0 : response.cursor) === ""),
                getItems: (response) => { var _a; return (_a = response === null || response === void 0 ? void 0 : response.transferOrders) !== null && _a !== void 0 ? _a : []; },
                loadPage: (response) => {
                    return list(core.setObjectProperty(request, "cursor", response === null || response === void 0 ? void 0 : response.cursor));
                },
            });
        });
    }
    /**
     * Retrieves a specific [TransferOrder](entity:TransferOrder) by ID. Returns the complete
     * order details including:
     *
     * - Basic information (status, dates, notes)
     * - Line items with ordered and received quantities
     * - Source and destination [Location](entity:Location)s
     * - Tracking information (if available)
     *
     * @param {Square.GetTransferOrdersRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.get({
     *         transferOrderId: "transfer_order_id"
     *     })
     */
    get(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get(request, requestOptions));
    }
    __get(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId } = request;
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}`),
                method: "GET",
                headers: _headers,
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.RetrieveTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "GET", "/v2/transfer-orders/{transfer_order_id}");
        });
    }
    /**
     * Updates an existing transfer order. This endpoint supports sparse updates,
     * allowing you to modify specific fields without affecting others.
     *
     * Creates a [transfer_order.updated](webhook:transfer_order.updated) webhook event.
     *
     * @param {Square.UpdateTransferOrderRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.update({
     *         transferOrderId: "transfer_order_id",
     *         idempotencyKey: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
     *         transferOrder: {
     *             sourceLocationId: "EXAMPLE_SOURCE_LOCATION_ID_789",
     *             destinationLocationId: "EXAMPLE_DEST_LOCATION_ID_101",
     *             expectedAt: "2025-11-10T08:00:00Z",
     *             notes: "Updated: Priority transfer due to low stock at destination",
     *             trackingNumber: "TRACK987654321",
     *             lineItems: [{
     *                     uid: "1",
     *                     quantityOrdered: "7"
     *                 }, {
     *                     itemVariationId: "EXAMPLE_NEW_ITEM_VARIATION_ID_003",
     *                     quantityOrdered: "2"
     *                 }, {
     *                     uid: "2",
     *                     remove: true
     *                 }]
     *         },
     *         version: BigInt("1753109537351")
     *     })
     */
    update(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__update(request, requestOptions));
    }
    __update(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId } = request, _body = __rest(request, ["transferOrderId"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}`),
                method: "PUT",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: serializers.UpdateTransferOrderRequest.jsonOrThrow(_body, {
                    unrecognizedObjectKeys: "strip",
                    omitUndefined: true,
                }),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.UpdateTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "PUT", "/v2/transfer-orders/{transfer_order_id}");
        });
    }
    /**
     * Deletes a transfer order in [DRAFT](entity:TransferOrderStatus) status.
     * Only draft orders can be deleted. Once an order is started via
     * [StartTransferOrder](api-endpoint:TransferOrders-StartTransferOrder), it can no longer be deleted.
     *
     * Creates a [transfer_order.deleted](webhook:transfer_order.deleted) webhook event.
     *
     * @param {Square.DeleteTransferOrdersRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.delete({
     *         transferOrderId: "transfer_order_id",
     *         version: BigInt("1000000")
     *     })
     */
    delete(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__delete(request, requestOptions));
    }
    __delete(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId, version } = request;
            const _queryParams = {
                version,
            };
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}`),
                method: "DELETE",
                headers: _headers,
                queryString: core.url
                    .queryBuilder()
                    .addMany(_queryParams)
                    .mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams)
                    .build(),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.DeleteTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "DELETE", "/v2/transfer-orders/{transfer_order_id}");
        });
    }
    /**
     * Cancels a transfer order in [STARTED](entity:TransferOrderStatus) or
     * [PARTIALLY_RECEIVED](entity:TransferOrderStatus) status. Any unreceived quantities will no
     * longer be receivable and will be immediately returned to the source [Location](entity:Location)'s inventory.
     *
     * Common reasons for cancellation:
     * - Items no longer needed at destination
     * - Source location needs the inventory
     * - Order created in error
     *
     * Creates a [transfer_order.updated](webhook:transfer_order.updated) webhook event.
     *
     * @param {Square.CancelTransferOrderRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.cancel({
     *         transferOrderId: "transfer_order_id",
     *         idempotencyKey: "65cc0586-3e82-4d08-b524-3885cffd52",
     *         version: BigInt("1753117449752")
     *     })
     */
    cancel(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__cancel(request, requestOptions));
    }
    __cancel(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId } = request, _body = __rest(request, ["transferOrderId"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}/cancel`),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: serializers.CancelTransferOrderRequest.jsonOrThrow(_body, {
                    unrecognizedObjectKeys: "strip",
                    omitUndefined: true,
                }),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.CancelTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v2/transfer-orders/{transfer_order_id}/cancel");
        });
    }
    /**
     * Records receipt of [CatalogItemVariation](entity:CatalogItemVariation)s for a transfer order.
     * This endpoint supports partial receiving - you can receive items in multiple batches.
     *
     * For each line item, you can specify:
     * - Quantity received in good condition (added to destination inventory with [InventoryState](entity:InventoryState) of IN_STOCK)
     * - Quantity damaged during transit/handling (added to destination inventory with [InventoryState](entity:InventoryState) of WASTE)
     * - Quantity canceled (returned to source location's inventory)
     *
     * The order must be in [STARTED](entity:TransferOrderStatus) or [PARTIALLY_RECEIVED](entity:TransferOrderStatus) status.
     * Received quantities are added to the destination [Location](entity:Location)'s inventory according to their condition.
     * Canceled quantities are immediately returned to the source [Location](entity:Location)'s inventory.
     *
     * When all items are either received, damaged, or canceled, the order moves to
     * [COMPLETED](entity:TransferOrderStatus) status.
     *
     * Creates a [transfer_order.updated](webhook:transfer_order.updated) webhook event.
     *
     * @param {Square.ReceiveTransferOrderRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.receive({
     *         transferOrderId: "transfer_order_id",
     *         idempotencyKey: "EXAMPLE_IDEMPOTENCY_KEY_101",
     *         receipt: {
     *             lineItems: [{
     *                     transferOrderLineUid: "1",
     *                     quantityReceived: "3",
     *                     quantityDamaged: "1",
     *                     quantityCanceled: "1"
     *                 }, {
     *                     transferOrderLineUid: "2",
     *                     quantityReceived: "2",
     *                     quantityCanceled: "1"
     *                 }]
     *         },
     *         version: BigInt("1753118664873")
     *     })
     */
    receive(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__receive(request, requestOptions));
    }
    __receive(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId } = request, _body = __rest(request, ["transferOrderId"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}/receive`),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: serializers.ReceiveTransferOrderRequest.jsonOrThrow(_body, {
                    unrecognizedObjectKeys: "strip",
                    omitUndefined: true,
                }),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.ReceiveTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v2/transfer-orders/{transfer_order_id}/receive");
        });
    }
    /**
     * Changes a [DRAFT](entity:TransferOrderStatus) transfer order to [STARTED](entity:TransferOrderStatus) status.
     * This decrements inventory at the source [Location](entity:Location) and marks it as in-transit.
     *
     * The order must be in [DRAFT](entity:TransferOrderStatus) status and have all required fields populated.
     * Once started, the order can no longer be deleted, but it can be canceled via
     * [CancelTransferOrder](api-endpoint:TransferOrders-CancelTransferOrder).
     *
     * Creates a [transfer_order.updated](webhook:transfer_order.updated) webhook event.
     *
     * @param {Square.StartTransferOrderRequest} request
     * @param {TransferOrdersClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.transferOrders.start({
     *         transferOrderId: "transfer_order_id",
     *         idempotencyKey: "EXAMPLE_IDEMPOTENCY_KEY_789",
     *         version: BigInt("1753109537351")
     *     })
     */
    start(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__start(request, requestOptions));
    }
    __start(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
            const { transferOrderId } = request, _body = __rest(request, ["transferOrderId"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, (0, headers_1.mergeOnlyDefinedHeaders)({ "Square-Version": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.version) !== null && _b !== void 0 ? _b : "2026-07-15" }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_c = this._options.fetcher) !== null && _c !== void 0 ? _c : core.fetcher)({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.SquareEnvironment.Production, `v2/transfer-orders/${core.url.encodePathParam(transferOrderId)}/start`),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: serializers.StartTransferOrderRequest.jsonOrThrow(_body, {
                    unrecognizedObjectKeys: "strip",
                    omitUndefined: true,
                }),
                timeoutMs: ((_h = (_f = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _f !== void 0 ? _f : (_g = this._options) === null || _g === void 0 ? void 0 : _g.timeoutInSeconds) !== null && _h !== void 0 ? _h : 60) * 1000,
                maxRetries: (_j = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _j !== void 0 ? _j : (_k = this._options) === null || _k === void 0 ? void 0 : _k.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_l = this._options) === null || _l === void 0 ? void 0 : _l.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: serializers.StartTransferOrderResponse.parseOrThrow(_response.body, {
                        unrecognizedObjectKeys: "passthrough",
                        allowUnrecognizedUnionMembers: true,
                        allowUnrecognizedEnumValues: true,
                        skipValidation: true,
                        breadcrumbsPrefix: ["response"],
                    }),
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                throw new errors.SquareError({
                    statusCode: _response.error.statusCode,
                    body: _response.error.body,
                    rawResponse: _response.rawResponse,
                });
            }
            return (0, handleNonStatusCodeError_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v2/transfer-orders/{transfer_order_id}/start");
        });
    }
}
exports.TransferOrdersClient = TransferOrdersClient;
