// square/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SquareClient = void 0;
const Client_1 = require("./api/resources/applePay/client/Client");
const Client_2 = require("./api/resources/bankAccounts/client/Client");
const Client_3 = require("./api/resources/bookings/client/Client");
const Client_4 = require("./api/resources/cards/client/Client");
const Client_5 = require("./api/resources/cashDrawers/client/Client");
const Client_6 = require("./api/resources/catalog/client/Client");
const Client_7 = require("./api/resources/channels/client/Client");
const Client_8 = require("./api/resources/checkout/client/Client");
const Client_9 = require("./api/resources/customers/client/Client");
const Client_10 = require("./api/resources/devices/client/Client");
const Client_11 = require("./api/resources/disputes/client/Client");
const Client_12 = require("./api/resources/employees/client/Client");
const Client_13 = require("./api/resources/events/client/Client");
const Client_14 = require("./api/resources/giftCards/client/Client");
const Client_15 = require("./api/resources/inventory/client/Client");
const Client_16 = require("./api/resources/invoices/client/Client");
const Client_17 = require("./api/resources/labor/client/Client");
const Client_18 = require("./api/resources/locations/client/Client");
const Client_19 = require("./api/resources/loyalty/client/Client");
const Client_20 = require("./api/resources/merchants/client/Client");
const Client_21 = require("./api/resources/oAuth/client/Client");
const Client_22 = require("./api/resources/orders/client/Client");
const Client_23 = require("./api/resources/payments/client/Client");
const Client_24 = require("./api/resources/payouts/client/Client");
const Client_25 = require("./api/resources/refunds/client/Client");
const Client_26 = require("./api/resources/reporting/client/Client");
const Client_27 = require("./api/resources/sites/client/Client");
const Client_28 = require("./api/resources/snippets/client/Client");
const Client_29 = require("./api/resources/subscriptions/client/Client");
const Client_30 = require("./api/resources/team/client/Client");
const Client_31 = require("./api/resources/teamMembers/client/Client");
const Client_32 = require("./api/resources/terminal/client/Client");
const Client_33 = require("./api/resources/transferOrders/client/Client");
const Client_34 = require("./api/resources/v1Transactions/client/Client");
const Client_35 = require("./api/resources/vendors/client/Client");
const Client_36 = require("./api/resources/webhooks/client/Client");
const BaseClient_1 = require("./BaseClient");
const core = __importStar(require("./core"));
class SquareClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_1.normalizeClientOptionsWithAuth)(options);
    }
    get oAuth() {
        var _a;
        return ((_a = this._oAuth) !== null && _a !== void 0 ? _a : (this._oAuth = new Client_21.OAuthClient(this._options)));
    }
    get v1Transactions() {
        var _a;
        return ((_a = this._v1Transactions) !== null && _a !== void 0 ? _a : (this._v1Transactions = new Client_34.V1TransactionsClient(this._options)));
    }
    get applePay() {
        var _a;
        return ((_a = this._applePay) !== null && _a !== void 0 ? _a : (this._applePay = new Client_1.ApplePayClient(this._options)));
    }
    get bankAccounts() {
        var _a;
        return ((_a = this._bankAccounts) !== null && _a !== void 0 ? _a : (this._bankAccounts = new Client_2.BankAccountsClient(this._options)));
    }
    get bookings() {
        var _a;
        return ((_a = this._bookings) !== null && _a !== void 0 ? _a : (this._bookings = new Client_3.BookingsClient(this._options)));
    }
    get cards() {
        var _a;
        return ((_a = this._cards) !== null && _a !== void 0 ? _a : (this._cards = new Client_4.CardsClient(this._options)));
    }
    get catalog() {
        var _a;
        return ((_a = this._catalog) !== null && _a !== void 0 ? _a : (this._catalog = new Client_6.CatalogClient(this._options)));
    }
    get channels() {
        var _a;
        return ((_a = this._channels) !== null && _a !== void 0 ? _a : (this._channels = new Client_7.ChannelsClient(this._options)));
    }
    get customers() {
        var _a;
        return ((_a = this._customers) !== null && _a !== void 0 ? _a : (this._customers = new Client_9.CustomersClient(this._options)));
    }
    get devices() {
        var _a;
        return ((_a = this._devices) !== null && _a !== void 0 ? _a : (this._devices = new Client_10.DevicesClient(this._options)));
    }
    get disputes() {
        var _a;
        return ((_a = this._disputes) !== null && _a !== void 0 ? _a : (this._disputes = new Client_11.DisputesClient(this._options)));
    }
    get employees() {
        var _a;
        return ((_a = this._employees) !== null && _a !== void 0 ? _a : (this._employees = new Client_12.EmployeesClient(this._options)));
    }
    get events() {
        var _a;
        return ((_a = this._events) !== null && _a !== void 0 ? _a : (this._events = new Client_13.EventsClient(this._options)));
    }
    get giftCards() {
        var _a;
        return ((_a = this._giftCards) !== null && _a !== void 0 ? _a : (this._giftCards = new Client_14.GiftCardsClient(this._options)));
    }
    get inventory() {
        var _a;
        return ((_a = this._inventory) !== null && _a !== void 0 ? _a : (this._inventory = new Client_15.InventoryClient(this._options)));
    }
    get invoices() {
        var _a;
        return ((_a = this._invoices) !== null && _a !== void 0 ? _a : (this._invoices = new Client_16.InvoicesClient(this._options)));
    }
    get labor() {
        var _a;
        return ((_a = this._labor) !== null && _a !== void 0 ? _a : (this._labor = new Client_17.LaborClient(this._options)));
    }
    get locations() {
        var _a;
        return ((_a = this._locations) !== null && _a !== void 0 ? _a : (this._locations = new Client_18.LocationsClient(this._options)));
    }
    get loyalty() {
        var _a;
        return ((_a = this._loyalty) !== null && _a !== void 0 ? _a : (this._loyalty = new Client_19.LoyaltyClient(this._options)));
    }
    get merchants() {
        var _a;
        return ((_a = this._merchants) !== null && _a !== void 0 ? _a : (this._merchants = new Client_20.MerchantsClient(this._options)));
    }
    get checkout() {
        var _a;
        return ((_a = this._checkout) !== null && _a !== void 0 ? _a : (this._checkout = new Client_8.CheckoutClient(this._options)));
    }
    get orders() {
        var _a;
        return ((_a = this._orders) !== null && _a !== void 0 ? _a : (this._orders = new Client_22.OrdersClient(this._options)));
    }
    get payments() {
        var _a;
        return ((_a = this._payments) !== null && _a !== void 0 ? _a : (this._payments = new Client_23.PaymentsClient(this._options)));
    }
    get payouts() {
        var _a;
        return ((_a = this._payouts) !== null && _a !== void 0 ? _a : (this._payouts = new Client_24.PayoutsClient(this._options)));
    }
    get refunds() {
        var _a;
        return ((_a = this._refunds) !== null && _a !== void 0 ? _a : (this._refunds = new Client_25.RefundsClient(this._options)));
    }
    get sites() {
        var _a;
        return ((_a = this._sites) !== null && _a !== void 0 ? _a : (this._sites = new Client_27.SitesClient(this._options)));
    }
    get snippets() {
        var _a;
        return ((_a = this._snippets) !== null && _a !== void 0 ? _a : (this._snippets = new Client_28.SnippetsClient(this._options)));
    }
    get subscriptions() {
        var _a;
        return ((_a = this._subscriptions) !== null && _a !== void 0 ? _a : (this._subscriptions = new Client_29.SubscriptionsClient(this._options)));
    }
    get teamMembers() {
        var _a;
        return ((_a = this._teamMembers) !== null && _a !== void 0 ? _a : (this._teamMembers = new Client_31.TeamMembersClient(this._options)));
    }
    get team() {
        var _a;
        return ((_a = this._team) !== null && _a !== void 0 ? _a : (this._team = new Client_30.TeamClient(this._options)));
    }
    get terminal() {
        var _a;
        return ((_a = this._terminal) !== null && _a !== void 0 ? _a : (this._terminal = new Client_32.TerminalClient(this._options)));
    }
    get transferOrders() {
        var _a;
        return ((_a = this._transferOrders) !== null && _a !== void 0 ? _a : (this._transferOrders = new Client_33.TransferOrdersClient(this._options)));
    }
    get vendors() {
        var _a;
        return ((_a = this._vendors) !== null && _a !== void 0 ? _a : (this._vendors = new Client_35.VendorsClient(this._options)));
    }
    get reporting() {
        var _a;
        return ((_a = this._reporting) !== null && _a !== void 0 ? _a : (this._reporting = new Client_26.ReportingClient(this._options)));
    }
    get cashDrawers() {
        var _a;
        return ((_a = this._cashDrawers) !== null && _a !== void 0 ? _a : (this._cashDrawers = new Client_5.CashDrawersClient(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new Client_36.WebhooksClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : this._options.environment,
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
exports.SquareClient = SquareClient;
