// square/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportingHelper = exports.WebhooksHelper = exports.serialization = exports.SquareTimeoutError = exports.SquareError = exports.SquareEnvironment = exports.SquareClient = exports.Square = void 0;
__exportStar(require("./api/index.js"), exports);
exports.Square = __importStar(require("./api/index.js"));
var Client_js_1 = require("./Client.js");
Object.defineProperty(exports, "SquareClient", { enumerable: true, get: function () { return Client_js_1.SquareClient; } });
var environments_js_1 = require("./environments.js");
Object.defineProperty(exports, "SquareEnvironment", { enumerable: true, get: function () { return environments_js_1.SquareEnvironment; } });
var index_js_1 = require("./errors/index.js");
Object.defineProperty(exports, "SquareError", { enumerable: true, get: function () { return index_js_1.SquareError; } });
Object.defineProperty(exports, "SquareTimeoutError", { enumerable: true, get: function () { return index_js_1.SquareTimeoutError; } });
__exportStar(require("./exports.js"), exports);
exports.serialization = __importStar(require("./serialization/index.js"));
var WebhooksHelper_1 = require("./wrapper/WebhooksHelper");
Object.defineProperty(exports, "WebhooksHelper", { enumerable: true, get: function () { return WebhooksHelper_1.WebhooksHelper; } });
var ReportingHelper_1 = require("./wrapper/ReportingHelper");
Object.defineProperty(exports, "ReportingHelper", { enumerable: true, get: function () { return ReportingHelper_1.ReportingHelper; } });
