// square/api/types/InventoryAdjustmentReasonIdType.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.InventoryAdjustmentReasonIdType = void 0;
/**
 * The type of inventory adjustment reason.
 *
 * Custom reasons use `CUSTOM` and include a `custom_reason_id`. Standard
 * seller-selectable reasons use `RECEIVED`, `DAMAGED`, `THEFT`, `LOST`,
 * `RETURNED`, `SPOILAGE_WASTE`, `SAMPLES_PROMOTIONAL`, `INTERNAL_USE`,
 * `VENDOR_RETURN`, or `PRODUCTION_WASTE`.
 *
 * The writable standard reasons with direction `INCREASE` are `RECEIVED`
 * and `RETURNED`. The writable standard reasons with direction `DECREASE`
 * are `DAMAGED`, `THEFT`, `LOST`, `SPOILAGE_WASTE`,
 * `SAMPLES_PROMOTIONAL`, `INTERNAL_USE`, `VENDOR_RETURN`, and
 * `PRODUCTION_WASTE`.
 *
 * The `SALE`, `RECOUNT`, `TRANSFER`, `IN_TRANSIT`, and `CANCELED_SALE`
 * values identify system-generated inventory events. These values can be
 * returned by Square, but cannot be used when writing inventory adjustments.
 */
exports.InventoryAdjustmentReasonIdType = {
    Received: "RECEIVED",
    Damaged: "DAMAGED",
    Theft: "THEFT",
    Lost: "LOST",
    Returned: "RETURNED",
    SpoilageWaste: "SPOILAGE_WASTE",
    SamplesPromotional: "SAMPLES_PROMOTIONAL",
    InternalUse: "INTERNAL_USE",
    VendorReturn: "VENDOR_RETURN",
    ProductionWaste: "PRODUCTION_WASTE",
    Sale: "SALE",
    Recount: "RECOUNT",
    Transfer: "TRANSFER",
    InTransit: "IN_TRANSIT",
    CanceledSale: "CANCELED_SALE",
    Custom: "CUSTOM",
};
